File: sliding_window/longest_repeating_character_replacement/js/longest_repeating_character_replacement.js
const s = "AABABBA";
const k = 1;

const characterReplacement = (s, k) => {
  let [l, r] = [0, 0];
  let index = [];
  let maxChar = 0;
  let best = 0;

  while(r < s.length) {
    let i = index[s[r]] || 0;
    index[s[r]] = ++i;
    maxChar = Math.max(maxChar, index[s[r]]);
    if(r - l + 1 - maxChar > k) {
      index[s[l]]--;
      l++;
    }
    best = Math.max(best, r - l + 1);
    r++;
  }
  return best;
}

console.log(characterReplacement(s, k));

File: sliding_window/longest_repeating_character_replacement/cpp/longest_repeating_character_replacement.cpp
#include<iostream>
#include<string>

using std::string;

// Time O(n)
// Space O(1)

class Solution {
public:
    static int characterReplacement(const string& s, int k) {
        int index[26]{};
        int l = 0, r = 0;
        int max_char = 0;
        int best = 0;

        /* We keep char index with visited chars
         * the max is the maximizing char that does 
         * not get counted against k
         * we then compare the maximizer against the substring
         * and if > k, pop back. Since the max will set
         * the first time it's hit we only need simple if block
         * about twice as fast as while loop
         */
        while(r < s.size()) {
            index[s[r]-65]++;
            max_char = std::max(max_char, index[s[r]-65]);
            if(r - l + 1 - max_char > k) {
                index[s[l]-65]--;
                l++;
            }
            best = std::max(best, r - l + 1);
            r++;
        }
        return best;
    }
};

int main() {
    string s = "AABABBA";
    int k = 1;

    int answer = Solution::characterReplacement(s, k);
    std::cout << answer << '\n';

    return 0;
}
